design/modular-versioning.md

```diff
@@ -0,0 +1,55 @@
+# Modular Versioning
+## Motivation
+The AWS Cloud Development Kit is a complex project composed of over `100` packages. The vast
+majority of those packages is made of construct libraries that expose each individual AWS services'
+surface area to developers. Those libraries need to be allowed to mature (at reach *GA*) at their
+own pace, in particular as the intention is to eventually transfer ownership of those to the service
+teams. This means CDK packages need to be versioned individually (in other words: in a modular way).
+Individual modules might also eventually be split out of the [awslabs/aws-cdk] mono-repo.
+
```

Comment:
One thing I think we need to deconstruct is the notion of pre- and post-1.0. Maybe we can live in a world that post-1.0, there is no such thing as a major version bump of an individual module. It will obvsiouly reduce our agility to a certain degree, but it will be much easier for users and maintainers to reason about what should work with what. Now the problem is pre-1.0 modules, which we should be able to major-version bump until it's ready for prime time...

My point is that this paragraph has some hidden assumptions about requirements and we should distill those very clearly prior to our session.

Reply:
I was somewhat intentionally avoiding to make a distinction between pre- and post-1.0 packages. I don't think there's any reason why they should be treated separately unless this dramatically simplifies the problem space (and I don't think it does).

Then, I would rather not limit major version bumping individual packages post-1.0, as this basically means any new major version will have to be coordinated across the entire surface (becomes challenging when packages get migrated to their own repositories... and anyway, that kind of coordination effort...?).

Reply:
There's a very big difference in mentality between pre- and post- 1.0 modules, exactly around breaking changes. The bar for a breaking change raises dramatically and we would probably do everything we can to avoid them after 1.0. And I do think it will dramatically simplify our maintenance story, but let's leave that for our discussion.